crypto/rsa/rsa_gen.c

```diff
@@ -85,62 +134,191 @@ static int rsa_builtin_keygen(RSA *rsa, int bits, BIGNUM *e_value,
+        for (i = 2; i < primes; i++) {
+            pinfo = rsa_multip_info_new();
+            if (pinfo == NULL)
+                goto err;
+            (void)sk_RSA_PRIME_INFO_push(prime_infos, pinfo);
+        }
+    }
+
     if (BN_copy(rsa->e, e_value) == NULL)
         goto err;
 
-    /* generate p and q */
-    for (;;) {
-        if (!BN_generate_prime_ex(rsa->p, bitsp, 0, NULL, NULL, cb))
-            goto err;
-        if (!BN_sub(r2, rsa->p, BN_value_one()))
-            goto err;
-        if (!BN_gcd(r1, r2, rsa->e, ctx))
-            goto err;
-        if (BN_is_one(r1))
-            break;
-        if (!BN_GENCB_call(cb, 2, n++))
+    /* generate p, q and other primes (if any) */
+    for (i = 0; i < primes; i++) {
+        adj = 0;
+        retries = 0;
+
+        if (i == 0) {
+            prime = rsa->p;
+        } else if (i == 1) {
+            prime = rsa->q;
+        } else {
+            pinfo = sk_RSA_PRIME_INFO_value(prime_infos, i - 2);
+            prime = pinfo->r;
+        }
+
+        for (;;) {
+ redo:
+            if (!BN_generate_prime_ex(prime, bitsr[i] + adj, 0, NULL, NULL, cb))
+                goto err;
+            /*
+             * prime should not be equal to p, q, r_3...
+             * (those primes prior to this one)
+             */
+            {
+                int j;
+
+                for (j = 0; j < i; j++) {
+                    BIGNUM *prev_prime;
+
+                    if (j == 0)
+                        prev_prime = rsa->p;
+                    else if (j == 1)
+                        prev_prime = rsa->q;
+                    else
+                        prev_prime = sk_RSA_PRIME_INFO_value(prime_infos,
+                                                             j - 2)->r;
+
+                    if (!BN_cmp(prime, prev_prime)) {
+                        goto redo;
+                    }
+                }
+            }
+            if (!BN_sub(r2, prime, BN_value_one()))
+                goto err;
+            if (!BN_gcd(r1, r2, rsa->e, ctx))
+                goto err;
+            if (BN_is_one(r1))
+                break;
+            if (!BN_GENCB_call(cb, 2, n++))
+                goto err;
+        }
+
+        bitse += bitsr[i];
+
+        /* calculate n immediately to see if it's sufficient */
+        if (i == 1) {
+            /* we get at least 2 primes */
+            if (!BN_mul(r1, rsa->p, rsa->q, ctx))
+                goto err;
+        } else if (i != 0) {
+            /* modulus n = p * q * r_3 * r_4 ... */
+            if (!BN_mul(r1, rsa->n, prime, ctx))
+                goto err;
+        } else {
+            /* i == 0, do nothing */
+            if (!BN_GENCB_call(cb, 3, i))
+                goto err;
+            continue;
+        }
+        /*
+         * if |r1|, product of factors so far, is not as long as expected
+         * (by checking the first 4 bits are less than 0x9 or greater than
+         * 0xF). If so, re-generate the last prime.
```

Comment:
"first bits" is ambiguous term (not to mention kind of explosive in sense that it has potential to evoke spirits of endian war), "most significant" is one that is appropriate. Also the statement is confusing in sense that 4-bit value can't be larger than 0xF.

Reply:
BTW, I'm actually uneasy about this check for 9. Original rationale is to merely obscure the fact that key is multi-prime (which should *not* be viewed as private information in sense that adversary can figure it out anyway). But question is if by insisting on specific bit pattern in most significant 4 bits you indirectly insist on specific pattern in less significant bits. Because in such case you might leak more information than you're trying to conceal. For example you're trying to conceal the fact that it's multi-prime, but indirectly reveal that it's specifically 4-prime. Well, amount of primes should *not* be viewed as private information either, what I'm rather trying to say is that argument in favour of check for 9 might contradict itself.

Reply:
I think "most significant" is also not really appropriate term either.
The product is scaled down by the expected number of bits - 4,
and that value can be less than 9 or even 8.  It can only exceed 15 if adj = 1.

Reply:
Well, initial objection was specifically about term "first bits". I mean what is "first bit", where do you start counting? "Most significant" is the only unambiguous way to describe the bit of interest here. Then yes, there is *second* ambiguity in comment, *how many* most significant bits are we looking at. As we indeed can end up with 4, 5, 3, and yes, "product [being] scaled down by the expected number of bits minus 4" eliminates both ambiguities...